### PR TITLE
FT: Use data && make check synchronous

### PR DESCRIPTION
**Pull request by alexandre-merle**
- Replace request.query by an object data to
  pass to the function
- Allow using another service than s3 in the
  signingKey
- Split the check and the backend call for the
  authentification

### CODE REVIEW

**Comment by DavidPineauScality** [comment]
There are some parts of this PR that I'm going to remove very soon.
Would it be possible to rework it so that it only introduces the useful bits ? I'll comment on the unnecessary bits.

**Comment by DavidPineauScality** [comment]
Note that the plan to rework the API is described in #102

**Comment by alexandre-merle** [comment]
Updated :) with an synchronous api :)

**Comment by alexandre-merle** [comment]
end to end running http://ci.ironmann.io/gh/scality/Integration/1835

**Comment by alexandre-merle** [comment]
see http://ci.ironmann.io/gh/scality/Integration/1837

File: lib/auth/auth.js [hunk 1]
```diff
@@ -10,19 +10,15 @@ const constructStringToSignV4 = require('./v4/constructStringToSign');
 const convertUTCtoISO8601 = require('./v4/timeUtils').convertUTCtoISO8601;
 const crypto = require('crypto');
 const vaultUtilities = require('./in_memory/vaultUtilities');
-
-
+let vault = require('./vault');
```
Comment (LaurenSpiegel) [inline lib/auth/auth.js]: Why require vault -- let's just set handler.

File: lib/auth/auth.js [hunk 2]
```diff
@@ -32,41 +28,64 @@ auth.doAuth = (request, log, cb, awsService) => {
...
         } else if (authHeader.startsWith('AWS4')) {
             log.debug('authenticating request with Auth V4 using headers');
-            authV4.headerAuthCheck.check(request, log, cb, awsService);
-        } else {
-            log.warn('missing authorization security header');
-            return cb(errors.MissingSecurityHeader);
+            return authV4.headerAuthCheck.check(request, log, awsService);
         }
-    } else if (request.query.Signature) {
+        log.warn('missing authorization security header');
```
Comment (rahulreddy) [inline lib/auth/auth.js]: This should be trace

File: lib/auth/auth.js [hunk 3]
```diff
@@ -32,41 +28,64 @@ auth.doAuth = (request, log, cb, awsService) => {
...
+                return cb(null, authInfo);
+            });
+    } else if (res.version === 4) {
+        res.data.log = log;
+        return vault.authenticateV4Request(res.data, cb, awsService);
+    } else if (res.data instanceof AuthInfo) {
+        return cb(null, res.data);
     }
-    return undefined;
+    log.error('Cannot found authentification method', {
```
Comment (rahulreddy) [inline lib/auth/auth.js]: This could be re-worded to `authentication method not found`

File: lib/auth/auth.js [hunk 4]
```diff
@@ -32,41 +28,64 @@ auth.doAuth = (request, log, cb, awsService) => {
...
+    }
+    // If no auth information is provided in request, then
+    // user is part of 'All Users Group' so send back this
+    // group as the canonicalID
+    log.trace('No authentication provided. User identified as public');
+    const authInfo = new AuthInfo({ canonicalID: constants.publicId });
+    return { err: null, data: authInfo };
+};
+
+auth.doAuth = (request, log, cb, awsService, data) => {
```
Comment (LaurenSpiegel) [inline lib/auth/auth.js]: This function is temporary, I assume?

File: lib/auth/auth.js [hunk 5]
```diff
@@ -32,41 +28,64 @@ auth.doAuth = (request, log, cb, awsService) => {
...
-    (request, data, accessKey, secretKeyValue) => {
+    (request, data, accessKey, secretKeyValue, awsService) => {
         Object.assign(request, { headers: {} });
         const amzDate = convertUTCtoISO8601(Date.now());
         // get date without time
         const scopeDate = amzDate.slice(0, amzDate.indexOf('T'));
         const signedHeaders = 'host;x-amz-date;x-amz-content-sha256';
         const region = 'us-east-1';
-        const service = 'iam';
+        const service = awsService || 'iam';
```
Comment (LaurenSpiegel) [inline lib/auth/auth.js]: I think this is against our style rules.

Comment (alexandre-merle) [inline lib/auth/auth.js]: It's not :) and it's more efficient for v8 in that way :) but i can change it :)

File: tests/unit/auth/v4/headerAuthCheck.js [hunk 6]
```diff
@@ -200,39 +187,19 @@ describe('v4 headerAuthCheck', () => {
...
-            done();
-        });
+        const res = headerAuthCheck(alteredRequest, log);
+        clock.uninstall();
+        assert.ifError(res.err);
+        done();
     });
 
-    it('should successfully authenticate', done => {
+    it('should successfully return information', done => {
```
Comment (LaurenSpiegel) [inline tests/unit/auth/v4/headerAuthCheck.js]: Do you still need the done?
It seems you are just testing that the auth version is returned.  Is that all we should be testing here?  If so, let's change the should statement to -- it should return the auth version type

File: tests/unit/auth/v4/queryAuthCheck.js [hunk 7]
```diff
@@ -171,23 +154,20 @@ describe('v4 queryAuthCheck', () => {
...
-            done();
-        });
+        const res = queryAuthCheck(alteredRequest, log, alteredRequest.query);
+        clock.uninstall();
+        assert.deepStrictEqual(res.err, errors.InvalidArgument);
+        done();
     });
 
-    it('should successfully authenticate', done => {
+    it('should successfully return informations', done => {
```
Comment (LaurenSpiegel) [inline tests/unit/auth/v4/queryAuthCheck.js]: Same comment as above.

File: tests/unit/auth/v4/headerAuthCheck.js (hunk 6 above)
Comment (alexandre-merle) [inline tests/unit/auth/v4/headerAuthCheck.js]: Yes, in fact checking there is no error is enought, but we still check the good work of this function, not really relate to the version

File: lib/auth/auth.js (hunk 4 above)
Comment (alexandre-merle) [inline lib/auth/auth.js]: yes :)

**Comment by alexandre-merle** [comment]
Updated :)

**Comment by rahulreddy** [comment]
👍

**Comment by alexandre-merle** [comment]
end to end running http://ci.ironmann.io/gh/scality/Integration/1842

**Comment by LaurenSpiegel** [comment]
👍

**Comment by alexandre-merle** [comment]
end to end passed, dropme removed :)

**Pull request merged**